[memory error] Heap Buffer Overflow when searching an absolute path

I was playing around with [AddressSanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer) and found out that a lot of the tests fail with various memory errors. Most are memory leaks, but one case I found is consistently reproducible (at least on Ubuntu 16.10 and Arch) and potentially dangerous: **`filename_filter()` overruns the `path` heap buffer when an absolute path is given on the command line.**

I believe that the comment *[base_path always ends with "/\0" while the path doesn't, so this is safe](https://github.com/ggreer/the_silver_searcher/blob/master/src/ignore.c#L303)* in `ignore.c` is inaccurate. `base_path` comes from the `realpath()` function, and according to my debug prints, it *does not* get the trailing `/` that you seem to be relying on, which ends up making `path_start` point to one byte beyond `path`'s nullbyte.

If you specify an absolute path with a trailing slash on the command line, `parse_options()` will move the nullbyte back a byte, so the `path` buffer has an extra unused byte allocated after the end of the string, which is enough to avoid the buffer overflow.

Running with `-D` for debugging will make it fail in a different place because a `log_debug` tries to print `path` before `path_ignore_search()` actually uses it.

See my log below. Disclaimer, you'll need [this patch](https://github.com/aswild/the_silver_searcher/commit/815d6975ab11c685a78379d9455f8ed14cb034fc) to be able to properly put `-O0` in `CFLAGS`, plus install `clang` and `llvm`. Symlink `/usr/local/bin/llvm-symbolizer` to `/usr/bin/llvm-symbolizer` to get the function names and line numbers in the output.

```
» allen@vm-ubuntu1610[sanitize] ~/the_silver_searcher
% ./build.sh CC=clang CFLAGS="-g -O0 -fsanitize=address"
---snip---
configure: creating ./config.status
config.status: creating Makefile
config.status: creating the_silver_searcher.spec
config.status: creating src/config.h
config.status: executing depfiles commands
  CC       src/ignore.o
  CC       src/log.o
  CC       src/print.o
  CC       src/options.o
  CC       src/print_w32.o
  CC       src/scandir.o
  CC       src/search.o
  CC       src/lang.o
  CC       src/util.o
  CC       src/decompress.o
  CC       src/main.o
src/decompress.c:52:31: warning: cast from 'const void *' to 'unsigned char *' drops const qualifier [-Wcast-qual]
    stream.next_in = (Bytef *)buf;
                              ^
1 warning generated.
  CCLD     ag

» allen@vm-ubuntu1610[sanitize] ~/the_silver_searcher
% mkdir example && cd example

» allen@vm-ubuntu1610[sanitize] ~/the_silver_searcher/example
% printf 'hello\n' >foo.txt

» allen@vm-ubuntu1610[sanitize] ~/the_silver_searcher/example
% pwd
/home/allen/the_silver_searcher/example

» allen@vm-ubuntu1610[sanitize] ~/the_silver_searcher/example
% ../ag hello `pwd`
=================================================================
==1073==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60400000dff8 at pc 0x0000004eab70 bp 0x7ffc64e722b0 sp 0x7ffc64e722a8
READ of size 1 at 0x60400000dff8 thread T0
    #0 0x4eab6f in path_ignore_search /home/allen/the_silver_searcher/src/ignore.c:211:33
    #1 0x4ea636 in filename_filter /home/allen/the_silver_searcher/src/ignore.c:344:13
    #2 0x4fe6e7 in ag_scandir /home/allen/the_silver_searcher/src/scandir.c:28:13
    #3 0x502e8c in search_dir /home/allen/the_silver_searcher/src/search.c:492:15
    #4 0x514cc1 in main /home/allen/the_silver_searcher/src/main.c:185:13
    #5 0x7f459ffd73f0 in __libc_start_main /build/glibc-jxM2Ev/glibc-2.24/csu/../csu/libc-start.c:291
    #6 0x418f79 in _start (/home/allen/the_silver_searcher/ag+0x418f79)

0x60400000dff8 is located 0 bytes to the right of 40-byte region [0x60400000dfd0,0x60400000dff8)
allocated by thread T0 here:
    #0 0x452ac0 in strdup (/home/allen/the_silver_searcher/ag+0x452ac0)
    #1 0x50cd84 in ag_strdup /home/allen/the_silver_searcher/src/util.c:40:17
    #2 0x4fabf8 in parse_options /home/allen/the_silver_searcher/src/options.c:786:20
    #3 0x513032 in main /home/allen/the_silver_searcher/src/main.c:51:5
    #4 0x7f459ffd73f0 in __libc_start_main /build/glibc-jxM2Ev/glibc-2.24/csu/../csu/libc-start.c:291

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/allen/the_silver_searcher/src/ignore.c:211:33 in path_ignore_search
Shadow bytes around the buggy address:
  0x0c087fff9ba0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9bb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9bc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9bd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9be0: fa fa fd fd fd fd fd fd fa fa 00 00 00 00 00 fa
=>0x0c087fff9bf0: fa fa 00 00 00 00 00 00 fa fa 00 00 00 00 00[fa]
  0x0c087fff9c00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1073==ABORTING
```

--- a/configure.ac
+++ b/configure.ac
@@ -10,6 +10,7 @@ AM_INIT_AUTOMAKE([no-define foreign subdir-objects])
 AC_PROG_CC
 AM_PROG_CC_C_O
 AC_PREREQ([2.59])
+AC_PROG_GREP
 
 m4_ifdef(
     [AM_SILENT_RULES],
@@ -24,7 +25,12 @@ AX_PTHREAD(
 )
 
 # Run CFLAGS="-pg" ./configure if you want debug symbols
-CFLAGS="$CFLAGS $PTHREAD_CFLAGS $PCRE_CFLAGS -Wall -Wextra -Wformat=2 -Wno-format-nonliteral -Wshadow -Wpointer-arith -Wcast-qual -Wmissing-prototypes -Wno-missing-braces -std=gnu89 -D_GNU_SOURCE -O2"
+if test "`$GREP -c '\b-O' >/dev/null ; echo $?`" = "0" ; then
+    CFLAGS="$CFLAGS -O2"
+fi
+
+CFLAGS="$CFLAGS $PTHREAD_CFLAGS -Wall -Wextra -Wformat=2 -Wno-format-nonliteral -Wshadow"
+CFLAGS="$CFLAGS -Wpointer-arith -Wcast-qual -Wmissing-prototypes -Wno-missing-braces -std=gnu89 -D_GNU_SOURCE"
 LDFLAGS="$LDFLAGS"
 
 case $host in

--- a/src/ignore.c
+++ b/src/ignore.c
@@ -295,15 +295,7 @@ int filename_filter(const char *path, const struct dirent *dir, void *baton) {
     }
 
     scandir_baton_t *scandir_baton = (scandir_baton_t *)baton;
-    const char *base_path = scandir_baton->base_path;
-    const size_t base_path_len = scandir_baton->base_path_len;
-    const char *path_start = path;
-
-    for (i = 0; base_path[i] == path[i] && i < base_path_len; i++) {
-        /* base_path always ends with "/\0" while path doesn't, so this is safe */
-        path_start = path + i + 2;
-    }
-    log_debug("path_start %s filename %s", path_start, filename);
+    const char *path_start = scandir_baton->path_start;
 
     const char *extension = strchr(filename, '.');
     if (extension) {

--- a/src/options.c
+++ b/src/options.c
@@ -199,6 +199,7 @@ void parse_options(int argc, char **argv, char **base_paths[], char **paths[]) {
     int ch;
     size_t i;
     int path_len = 0;
+    int base_path_len = 0;
     int useless = 0;
     int group = 1;
     int help = 0;
@@ -775,6 +776,7 @@ void parse_options(int argc, char **argv, char **base_paths[], char **paths[]) {
     }
 
     char *path = NULL;
+    char *base_path = NULL;
 #ifdef PATH_MAX
     char *tmp = NULL;
 #endif
@@ -792,10 +794,20 @@ void parse_options(int argc, char **argv, char **base_paths[], char **paths[]) {
             (*paths)[i] = path;
 #ifdef PATH_MAX
             tmp = ag_malloc(PATH_MAX);
-            (*base_paths)[i] = realpath(path, tmp);
+            base_path = realpath(path, tmp);
 #else
-            (*base_paths)[i] = realpath(path, NULL);
+            base_path = realpath(path, NULL);
 #endif
+            if (base_path) {
+                base_path_len = strlen(base_path);
+                /* add trailing slash */
+                if (base_path_len > 1 && base_path[base_path_len - 1] != '/') {
+                    base_path = ag_realloc(base_path, base_path_len + 2);
+                    base_path[base_path_len] = '/';
+                    base_path[base_path_len + 1] = '\0';
+                }
+            }
+            (*base_paths)[i] = base_path;
         }
         /* Make sure we search these paths instead of stdin. */
         opts.search_stream = 0;

--- a/src/scandir.h
+++ b/src/scandir.h
@@ -7,6 +7,7 @@ typedef struct {
     const ignores *ig;
     const char *base_path;
     size_t base_path_len;
+    const char *path_start;
 } scandir_baton_t;
 
 typedef int (*filter_fp)(const char *path, const struct dirent *, void *);

--- a/src/search.c
+++ b/src/search.c
@@ -463,6 +463,8 @@ void search_dir(ignores *ig, const char *base_path, const char *path, const int 
     struct dirent *dir = NULL;
     scandir_baton_t scandir_baton;
     int results = 0;
+    size_t base_path_len = 0;
+    const char *path_start = path;
 
     char *dir_full_path = NULL;
     const char *ignore_file = NULL;
@@ -486,9 +488,20 @@ void search_dir(ignores *ig, const char *base_path, const char *path, const int 
         dir_full_path = NULL;
     }
 
+    /* path_start is the part of path that isn't in base_path
+     * base_path will have a trailing '/' because we put it there in parse_options
+     */
+    base_path_len = base_path ? strlen(base_path) : 0;
+    for (i = 0; ((size_t)i < base_path_len) && (path[i]) && (base_path[i] == path[i]); i++) {
+        path_start = path + i + 1;
+    }
+    log_debug("search_dir: path is '%s', base_path is '%s', path_start is '%s'", path, base_path, path_start);
+
     scandir_baton.ig = ig;
     scandir_baton.base_path = base_path;
-    scandir_baton.base_path_len = base_path ? strlen(base_path) : 0;
+    scandir_baton.base_path_len = base_path_len;
+    scandir_baton.path_start = path_start;
+
     results = ag_scandir(path, &dir_list, &filename_filter, &scandir_baton);
     if (results == 0) {
         log_debug("No results found in directory %s", path);

